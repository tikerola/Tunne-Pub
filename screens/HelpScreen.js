import { Entypo, Ionicons } from "@expo/vector-icons";
import AsyncStorage from "@react-native-community/async-storage";
import { useFocusEffect } from "@react-navigation/native";
import React, { useEffect } from "react";
import { Dimensions, ScrollView, StyleSheet, Text, View } from "react-native";
import { useDispatch, useSelector } from "react-redux";
import { colors } from "../constants/colors";
import { dimensions } from "../constants/dimensions";
import {
  loadSummaryItems,
  setActiveSummaryItemId,
} from "../store/actions/cards";

const HelpScreen = () => {
  const dispatch = useDispatch();
  const { height, width } = Dimensions.get("window");

  const item = useSelector(
    (state) => state.cards.summaryOfChosenItems[state.cards.activeSummaryItemId]
  );

  // Make sure that if we are editing a record and go to help screen, we terminate the editing and
  // prepare for the possible creation of new session

  useFocusEffect(
    React.useCallback(() => {
      if (item && item.itemSaved) {
        dispatch(setActiveSummaryItemId(""));
      }
    }, [item])
  );

  // Fetch and initiate redux with async storage items

  useEffect(() => {
    const getFromAsyncStorage = async () => {
      try {
        const data = await AsyncStorage.getItem("items");
        const parsedData = JSON.parse(data);

        if (parsedData) {
          dispatch(loadSummaryItems(parsedData));
        }
      } catch (err) {
        console.log("Something went wrong when trying to fetch items");
      }
    };

    getFromAsyncStorage();
  }, []);

  return (
    <View style={styles.container}>
      <ScrollView contentContainerStyle={styles.contentContainer}>
        <View
          style={{
            width: width <= dimensions.phoneMaxWidth ? "100%" : "75%",
            alignSelf: "center",
          }}
        >
          <Text
            testID="title"
            style={[
              styles.title,
              { marginTop: height <= dimensions.phoneMaxHeight ? 20 : 50 },
            ]}
          >
            Tervetuloa
          </Text>

          <Text style={styles.content}>
            Tunne-sovelluksessa pääset luomaan uuden tunnedokumentin painamalla
            alavalikon keskimmäistä painiketta{" "}
            <Entypo name={"new-message"} size={20} color={"gray"} />. Jos olet
            tekemässä ensimmäistä istuntoa tai sinulla ei ole keskeneräisiä
            istuntoja, pyytää ohjelma sinua antamaan istunnolle nimi.
          </Text>

          <Text style={styles.content}>
            Tämän jälkeen eteesi avautuu ensimmäinen kysymys. Kysymykseen
            vastataan kirjoittamalla vastaus tekstikenttään ja painamalla
            'tallenna'-nappia. Tallennettuasi vastauksen, ohjelma siirtää sinut
            automaattisesti seuraavaan kysymykseen. Jos kuitenkin vastaaminen
            tuntuu vaikealta ja haluat kenties palata kysymyksen pariin vasta
            myöhemmin, voit siirtyä 'seuraava'-linkillä seuraavaan kysymykseen.
          </Text>

          <Text style={styles.content}>
            Toisinaan kysymyksiin (kysymykset: 2, 6, 7, 8 ja 9) on mahdollista
            vastata joko valitsemalla listasta valmiita sanoja/lauseita tai
            vaihtoehtoisesti kirjoittaa oma vastaus. Voit vaihtaa valintaasi
            kysymyskentän alla olevasta moodi-kytkimestä. Kytkimen tekstistä
            sekä tekstikentän vihjeestä näet miten vastata.
          </Text>

          <Text style={styles.content}>
            Kysymyksiä on kaiken kaikkiaan kymmenen. Jos haluat nähdä, mitä olet
            vastannut aiempiin kysymyksiin, näet koosteen vastauksistasi
            painamalla oikean yläkulman{" "}
            <Ionicons name={"md-clipboard"} size={20} color={"gray"} /> ikonia.
            Koostesivulta pääset jatkamaan istunnon päivittämistä painamalla
            yläpalkin takaisin-nuolta.
          </Text>

          <Text style={styles.content}>
            Tallennat istunnon painaessasi kymmenennen kysymyksen
            tallenna-nappia/linkkiä. Tämän jälkeen istunto on tallennettu
            puhelimen muistiin. Voit tätä ja tulevia tallennettuja istuntoja
            tutkailla ja päivitellä oikean alareunan{" "}
            <Entypo name="save" size={20} color="gray" /> napista ja
            valitsemalla listasta haluamasi istunto. Jos myöhemmin muokkaat jo
            tallennettua istuntoa, voit vastata vain haluamiisi kysymyksiin ja
            tallentaa nämä. Tallennus astuu heti voimaan tarvitsematta käydä
            kaikkia kymmentä kysymystä läpi.
          </Text>

          <Text style={styles.content}>
            Jos jotain jäi epäselväksi, voi ohjeisiin (
            <Entypo name="help" size={20} color="gray" />
            -nappi vasemmalla alhaalla) palata missä vaiheessa tahansa ilman,
            että istunnon jo kirjoitettuja tietoja menetetään.
          </Text>
        </View>
      </ScrollView>
    </View>
  );
};

export default HelpScreen;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: colors.bodyBg,
  },

  contentContainer: {
    display: "flex",
    flexDirection: "column",
    justifyContent: "space-around",
  },
  title: {
    width: "100%",
    marginTop: 20,
    textAlign: "center",
    marginBottom: 0,
    fontSize: 19,
    fontFamily: "Roboto-Bold",
    color: "#bbb",
  },

  content: {
    margin: 10,
    color: "#999",
    fontSize: 18,
  },
});
