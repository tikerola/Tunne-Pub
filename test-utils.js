import React from "react";
import { NavigationContext } from "@react-navigation/native";
import { render } from "@testing-library/react-native";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import CARD_DATA from "./data/cards";
import NEEDS_DATA from "./data/needs";
import { SENTENCES_DATA } from "./data/sentences";
import WORD_DATA from "./data/words";

const INITIAL_STATE = {
  cards: {
    // cards: CARD_DATA,
    // words: WORD_DATA,
    // needs: NEEDS_DATA,
    // sentences: SENTENCES_DATA,
    activeSummaryItemId: "123",
    summaryOfChosenItems: {
      123: {
        id: "123",
        name: "",
        cards: [],
        feelingsWords: [],
        feelingsYouMissWords: [],
        needs: ["hei"],
        needsBehindAFeeling: [],
        sentences: [],
        questions: {},
        itemSaved: true,
      },
    },
    screenNumber: 1,
    numOfItems: 0,
    initialLoadFinished: false,
  },
};

const mockStore = configureStore([]);
const store = mockStore(INITIAL_STATE);

const navContext = {
  isFocused: () => true,
  // addListener returns an unscubscribe function.
  addListener: jest.fn(() => jest.fn()),
};

const reduxRender = (ui, options, storeData = store) =>
  render(ui, {
    wrapper: ({ children }) => (
      <Provider store={storeData}>
        <NavigationContext.Provider value={navContext}>
          {children}
        </NavigationContext.Provider>
      </Provider>
    ),
    ...options,
  });

// re-export everything for convenience
export * from "@testing-library/react-native";

// override render method
export { reduxRender as render, store };
