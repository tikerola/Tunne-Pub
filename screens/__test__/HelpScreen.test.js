import React from 'react'
import HelpScreen from '../HelpScreen'
import { render, cleanup, store, waitFor } from '../../test-utils'



beforeEach(() => {
  cleanup()
  store.clearActions()
})

describe('<HelpScreen />', () => {
  it('should render a title', () => {
    const { getByTestId } = render(<HelpScreen />)
    const title = getByTestId('title')
    expect(title.children).toContain("Tervetuloa")
  })

  it('should match snapshot', () => {
    const result = render(<HelpScreen />).toJSON()
    expect(result).toMatchSnapshot()
  })

  it('should dispatch reset activeSummaryItemId if coming to helpscreen during editing', async () => {

    const rendered = render(<HelpScreen />)
    const actions = await waitFor(() => store.getActions(), { timeout: 5000 })
    
    expect(actions[0].type).toEqual('SET ACTIVE SUMMARY ITEM ID')
  })

  it('should load items when asyncstorage is not empty', async () => {

    const renderer = render(<HelpScreen />)
    const actions = await waitFor(() => store.getActions(), { timeout: 5000 })
    

    expect(actions[1].type).toEqual('LOAD SUMMARY ITEMS')

  })

})
