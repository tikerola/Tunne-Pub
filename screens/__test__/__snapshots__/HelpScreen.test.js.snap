// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HelpScreen /> should match snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#5f0c38",
      "flex": 1,
    }
  }
>
  <RCTScrollView
    contentContainerStyle={
      Object {
        "display": "flex",
        "flexDirection": "column",
        "justifyContent": "space-around",
      }
    }
  >
    <View>
      <View
        style={
          Object {
            "alignSelf": "center",
            "width": "75%",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#bbb",
                "fontFamily": "Roboto-Bold",
                "fontSize": 19,
                "marginBottom": 0,
                "marginTop": 20,
                "textAlign": "center",
                "width": "100%",
              },
              Object {
                "marginTop": 50,
              },
            ]
          }
          testID="title"
        >
          Tervetuloa
        </Text>
        <Text
          style={
            Object {
              "color": "#999",
              "fontSize": 18,
              "margin": 10,
            }
          }
        >
          Tunne-sovelluksessa pääset luomaan uuden tunnedokumentin painamalla alavalikon keskimmäistä painiketta
           
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": "gray",
                  "fontSize": 20,
                },
                undefined,
                Object {
                  "fontFamily": "entypo",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            
          </Text>
          . Jos olet tekemässä ensimmäistä istuntoa tai sinulla ei ole keskeneräisiä istuntoja, pyytää ohjelma sinua antamaan istunnolle nimi.
        </Text>
        <Text
          style={
            Object {
              "color": "#999",
              "fontSize": 18,
              "margin": 10,
            }
          }
        >
          Tämän jälkeen eteesi avautuu ensimmäinen kysymys. Kysymykseen vastataan kirjoittamalla vastaus tekstikenttään ja painamalla 'tallenna'-nappia. Tallennettuasi vastauksen, ohjelma siirtää sinut automaattisesti seuraavaan kysymykseen. Jos kuitenkin vastaaminen tuntuu vaikealta ja haluat kenties palata kysymyksen pariin vasta myöhemmin, voit siirtyä 'seuraava'-linkillä seuraavaan kysymykseen.
        </Text>
        <Text
          style={
            Object {
              "color": "#999",
              "fontSize": 18,
              "margin": 10,
            }
          }
        >
          Toisinaan kysymyksiin (kysymykset: 2, 6, 7, 8 ja 9) on mahdollista vastata joko valitsemalla listasta valmiita sanoja/lauseita tai vaihtoehtoisesti kirjoittaa oma vastaus. Voit vaihtaa valintaasi kysymyskentän alla olevasta moodi-kytkimestä. Kytkimen tekstistä sekä tekstikentän vihjeestä näet miten vastata.
        </Text>
        <Text
          style={
            Object {
              "color": "#999",
              "fontSize": 18,
              "margin": 10,
            }
          }
        >
          Kysymyksiä on kaiken kaikkiaan kymmenen. Jos haluat nähdä, mitä olet vastannut aiempiin kysymyksiin, näet koosteen vastauksistasi painamalla oikean yläkulman
           
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": "gray",
                  "fontSize": 20,
                },
                undefined,
                Object {
                  "fontFamily": "ionicons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            
          </Text>
           ikonia. Koostesivulta pääset jatkamaan istunnon päivittämistä painamalla yläpalkin takaisin-nuolta.
        </Text>
        <Text
          style={
            Object {
              "color": "#999",
              "fontSize": 18,
              "margin": 10,
            }
          }
        >
          Tallennat istunnon painaessasi kymmenennen kysymyksen tallenna-nappia/linkkiä. Tämän jälkeen istunto on tallennettu puhelimen muistiin. Voit tätä ja tulevia tallennettuja istuntoja tutkailla ja päivitellä oikean alareunan
           
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": "gray",
                  "fontSize": 20,
                },
                undefined,
                Object {
                  "fontFamily": "entypo",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            
          </Text>
           napista ja valitsemalla listasta haluamasi istunto. Jos myöhemmin muokkaat jo tallennettua istuntoa, voit vastata vain haluamiisi kysymyksiin ja tallentaa nämä. Tallennus astuu heti voimaan tarvitsematta käydä kaikkia kymmentä kysymystä läpi.
        </Text>
        <Text
          style={
            Object {
              "color": "#999",
              "fontSize": 18,
              "margin": 10,
            }
          }
        >
          Jos jotain jäi epäselväksi, voi ohjeisiin (
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": "gray",
                  "fontSize": 20,
                },
                undefined,
                Object {
                  "fontFamily": "entypo",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            
          </Text>
          -nappi vasemmalla alhaalla) palata missä vaiheessa tahansa ilman, että istunnon jo kirjoitettuja tietoja menetetään.
        </Text>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
